internal/controllers/mic_reconciler.go

```diff
@@ -0,0 +1,336 @@
+			}
+			_, err := controllerutil.CreateOrPatch(ctx, mrhi.client, mbscObj, func() error {
+				mrhi.mbscHelper.SetModuleImageSpec(mbscObj, &imageSpec)
+				mbscObj.Spec.ImageRepoSecret = micObj.Spec.ImageRepoSecret
+				return controllerutil.SetOwnerReference(micObj, mbscObj, mrhi.scheme)
+			})
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+//go:generate mockgen -source=mic_reconciler.go -package=controllers -destination=mock_mic_reconciler.go pullPodManager
+type pullPodManager interface {
+	listImagesPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) ([]v1.Pod, error)
+	deletePod(ctx context.Context, pod *v1.Pod) error
+	createPullPod(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec, micObj *kmmv1beta1.ModuleImagesConfig) error
+	getPullPodForImage(pods []v1.Pod, image string) *v1.Pod
+}
+
+type pullPodManagerImpl struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+func newPullPodManager(client client.Client, scheme *runtime.Scheme) pullPodManager {
+	return &pullPodManagerImpl{
+		client: client,
+		scheme: scheme,
+	}
+}
+
+func (ppmi *pullPodManagerImpl) listImagesPullPods(ctx context.Context, micObj *kmmv1beta1.ModuleImagesConfig) ([]v1.Pod, error) {
+	logger := ctrl.LoggerFrom(ctx).WithValues("mic name", micObj.Name)
+
+	pl := v1.PodList{}
+
+	hl := client.HasLabels{imageLabelKey}
+	ml := client.MatchingLabels{moduleImageLabelKey: micObj.Name}
+
+	logger.V(1).Info("Listing mic image Pods")
+
+	if err := ppmi.client.List(ctx, &pl, client.InNamespace(micObj.Namespace), hl, ml); err != nil {
+		return nil, fmt.Errorf("could not list mic image pods for mic %s: %v", micObj.Name, err)
+	}
+
+	return pl.Items, nil
+}
+
+func (ppmi *pullPodManagerImpl) deletePod(ctx context.Context, pod *v1.Pod) error {
+	logger := ctrl.LoggerFrom(ctx)
+	if pod.DeletionTimestamp != nil {
+		logger.Info("DeletionTimestamp set, pod is already in deletion", "pod", pod.Name)
+		return nil
+	}
+	if err := ppmi.client.Delete(ctx, pod); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("failed to delete pull pod %s/%s: %v", pod.Namespace, pod.Name, err)
+	}
+	return nil
+}
+
+func (ppmi *pullPodManagerImpl) createPullPod(ctx context.Context, imageSpec *kmmv1beta1.ModuleImageSpec, micObj *kmmv1beta1.ModuleImagesConfig) error {
+	restartPolicy := v1.RestartPolicyOnFailure
+	if imageSpec.Build != nil || imageSpec.Sign != nil {
+		restartPolicy = v1.RestartPolicyNever
+	}
+	imagePullSecrets := []v1.LocalObjectReference{}
+	if micObj.Spec.ImageRepoSecret != nil {
+		imagePullSecrets = []v1.LocalObjectReference{*micObj.Spec.ImageRepoSecret}
+	}
+
+	pullPod := v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: micObj.Name + "-pull-pod-",
+			Namespace:    micObj.Namespace,
+			Labels: map[string]string{
+				moduleImageLabelKey: micObj.Name,
+				imageLabelKey:       imageSpec.Image,
+			},
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    workerContainerName,
```

Comment:
Why do we call this container `worker`? Isn't it confusing with the actual worker pods?

Reply:
fixed